### ISSUE TITLE
 undefined method `downcase' error on a builtin win10 firewall rule

### ISSUE THREAD

**Issue by kreeuwijk** [GeoffWilliams/puppet-windows_firewall#13]
On the latest Windows 10 version, ps-bridge.ps1 is generating a strange `RemotePort` value for 1 firewall rule, which trips up Ruby and stops the puppet run dead in its tracks with the following error:

```
Error: Could not prefetch windows_firewall_group provider 'windows_firewall_group': undefined method `downcase' for {"value"=>["554", "8554-8558"], "Count"=>2}:Hash
Error: Failed to apply catalog: undefined method `downcase' for {"value"=>["554", "8554-8558"], "Count"=>2}:Hash
```

Capturing the output via `ps-bridge.ps1 show` shows the initial rule with the strange RemotePort value:
```
{
        "Action":  "Allow",
        "Description":  "Inbound rule for the SNMP Trap Service to allow SNMP traps. [UDP 162]",
        "Direction":  "Inbound",
        "Name":  "SNMP Trap Service (UDP In)",
        "EdgeTraversalPolicy":  "Block",
        "Unused_RuleSource":  "Local Setting",
        "RemotePort":  "Any",
        "LocalAddress":  "Any",
        "Protocol":  "UDP",
        "LocalPort":  "162",
        "DisplayGroup":  "SNMP Trap",
        "Service":  "SNMPTRAP",
        "InterfaceType":  "Any",
        "IcmpType":  null,
        "Unused_Security":  "NotRequired",
        "DisplayName":  "SNMP Trap Service (UDP In)",
        "Profile":  "Private,Public",
        "Program":  "C:\\\\Windows\\\\system32\\\\snmptrap.exe",
        "RemoteAddress":  "LocalSubnet",
        "Enabled":  "False"
    },
    {
        "RemotePort":  {
                           "value":  [
                                         "554",
                                         "8554-8558"
                                     ],
                           "Count":  2
                       },
        "InterfaceType":  "Any",
        "Enabled":  "True",
        "LocalPort":  "Any",
        "Profile":  "Any",
        "DisplayGroup":  "@FirewallAPI.dll,-80200",
        "Description":  "@FirewallAPI.dll,-80203",
        "Program":  "%SystemRoot%\\system32\\svchost.exe",
        "Protocol":  "TCP",
        "LocalAddress":  "Any",
        "EdgeTraversalPolicy":  "Block",
        "Service":  "FrameServer",
        "Direction":  "Outbound",
        "RemoteAddress":  "LocalSubnet",
        "Action":  "Allow",
        "DisplayName":  "@FirewallAPI.dll,-80204",
        "IcmpType":  "Any",
        "Name":  "NVS-FrameServer-Out-TCP-NoScope"
    },
```

Removing the firewall rule does not solve the problem, as the strange RemotePort value then shows up in a different rule. This might be a Powershell bug. The version of Powershell is 5.1.18362.145

The code should be updated to check for the structure of the value, before deciding to run `.downcase` on it.

**Comment by kreeuwijk** [comment]
This problem does not occur on Windows Server 2012R2 or 2016. I have not tested on Windows Server 2019 yet, but it is more likely to occur there as that OS has more in common with Windows 10.

**Comment by GeoffWilliams** [comment]
Thanks for the analysis. I had a go trying to reproduce this this morning before work but ran out of time while windows was doing something with its updates. I am able to create a single rule combining a port+range and parse it with the `puppet resource`. It's very strange that deleting the 'bad' rule doesn't fix this and if there is a bug in powershell there's really not much that can be done to fix it. Could I please get you to have a go attaching a stack trace to this ticket? should be able to see it with `puppet apply yourfile.pp --debug --evaltrace` or `puppet agent -t --debug --evaltrace`.  Also please let me know the exact windows version and powershell versions your using so that I can tell my system is the same as yours. Alternatively if you want to have a go fixing this yourself patches are very appreciated. I'm working on other projects right now so my time on this is limited

**Comment by GeoffWilliams** [comment]
Scratch that - updates completed and bug reproduced. I'll see if I can come up with a quick fix

**Comment by GeoffWilliams** [comment]
There is a difference in the JSON output by the ps-bridge script on new windows:

```json
   {
        "RemotePort":  "Any",
        "InterfaceType":  "Any",
        "Enabled":  "True",
        "LocalPort":  {
                          "value":  [
                                        "554",
                                        "8554-8558"
                                    ],
                          "Count":  2
                      },
        "Profile":  "Any",
        "DisplayGroup":  "@FirewallAPI.dll,-80200",
        "Description":  "@FirewallAPI.dll,-80202",
        "Program":  "%SystemRoot%\\system32\\svchost.exe",
        "Protocol":  "TCP",
        "LocalAddress":  "Any",
        "EdgeTraversalPolicy":  "Block",
        "Service":  "FrameServer",
        "Direction":  "Inbound",
        "RemoteAddress":  "LocalSubnet",
        "Action":  "Allow",
        "DisplayName":  "@FirewallAPI.dll,-80201",
        "IcmpType":  "Any",
        "Name":  "NVS-FrameServer-In-TCP-NoScope"
    }
```

vs older windows (closest match):
```json
    {
        "Action":  "Allow",
        "Description":  "Inbound rule for Media Center Extenders to allow RTSP AV Control. [TCP 554, TCP 8554-8558]",
        "Direction":  "Inbound",
        "Name":  "Media Center Extenders - RTSP (TCP-In)",
        "EdgeTraversalPolicy":  "Block",
        "Unused_RuleSource":  "Local Setting",
        "RemotePort":  "Any",
        "LocalAddress":  "Any",
        "Protocol":  "TCP",
        "LocalPort":  "554,8554,8555,8556,8557,8558",
        "DisplayGroup":  "Media Center Extenders",
        "InterfaceType":  "Any",
        "IcmpType":  null,
        "Unused_Security":  "NotRequired",
        "DisplayName":  "Media Center Extenders - RTSP (TCP-In)",
        "Profile":  "Domain,Private,Public",
        "Program":  "C:\\\\Windows\\\\ehome\\\\ehshell.exe",
        "RemoteAddress":  "LocalSubnet",
        "Enabled":  "False"
    },

```

These are both from different patch levels of windows 10 which is what I use for all my testing/development. 

Working: windows build 17134.rs4_release.180410-1804, powershell 5.1.17134.1
Not working: windows build 18362.295 1903, powershell 5.1.118362.145

To fix this, the script needs to give the same format output on both platforms - eg value of `RemotePort` and `LocalPort` must not be hashes

**Comment by GeoffWilliams** [comment]
@kreeuwijk I seem to have fixed this. Please test the module in the `psmunt` branch of this repository that contains an updated script and let me know how you get on

**Comment by canihavethisone** [comment]
on Win10.1909 the proposed changes progress the puppet run, however throw an error for each rule removed, ie:
    `Could` not evaluate: Execution of 'powershell.exe -ExecutionPolicy Bypass -File C:/ProgramData/PuppetLabs/puppet/cache/lib/ps/windows_firewall/ps-bridge.ps1 delete Windows Media Player Network Sharing Service (UDP-Out)' returned 1: Deleting Windows Media Player Network Sharing Service (UDP-Out)...
We were told to delete firewall rule 'Windows Media Player Network Sharing Service (UDP-Out)' but it does not exist
At C:\ProgramData\PuppetLabs\puppet\cache\lib\ps\windows_firewall\ps-bridge.ps1:344 char:9
+ throw "We were told to delete firewall rule '$($name)' but it ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo : OperationStopped: (We were told to... does not exist:String) [], RuntimeException
+ FullyQualifiedErrorId : We were told to delete firewall rule 'Windows Media Player Network Sharing Service (UDP-
Out)' but it does not exist'

**Comment by canihavethisone** [comment]
@GeoffWilliams is this patch going to get merged?

**Comment by GeoffWilliams** [comment]
never got a reply on testing but I'm going to just assume it worked - have merged to master. LMK if anyone wants a forge release

**Comment by canihavethisone** [comment]
@GeoffWilliams yes please release it to the forge with updated metadata and changelog - it seems to work in my testing with 1903 & 1909.

BTW, I am not seeing it merged into your master branch.

**Comment by canihavethisone** [comment]
@GeoffWilliams FYI, I grew weary of the unavoidable 20+ mins to run Puppet the first time when purging undefined rules, so as I use Foreman with the Wimaging templates (soon to be merged into the mainstream release as Windows templates), I have added the block below to extraFinishCommands snippet which is called by the Finish template .  This does a very quick purge during the build so the first Puppet run is much faster.

I am curious if there is a way to do a bulk purge with an 'except' entry in your module.....

    <% if host_param_true?('windows-purge-firewall') -%>
      echo Purging firewall rules
      powershell.exe Remove-NetFirewallRule -All -Verbose
      powershell.exe Remove-ItemProperty -Path HKLM:\SYSTEM\CurrentControlSet\Services\SharedAccess\Parameters\FirewallPolicy\RestrictedServices\AppIso\FirewallRules -Name *
      powershell.exe Remove-ItemProperty -Path HKLM:\SYSTEM\CurrentControlSet\Services\SharedAccess\Parameters\FirewallPolicy\RestrictedServices\Configurable\System -Name *
      powershell.exe New-NetFirewallRule -DisplayName 'Puppet Agent (temporary)' -Enabled True -Action Allow -Profile ANY -Direction Outbound -Protocol TCP  -RemotePort 8140 -Program 'C:\Program Files\Puppet Labs\Puppet\sys\ruby\bin\ruby.exe'
      powershell.exe New-NetFirewallRule -DisplayName 'Internet Browsing (temporary)' -Enabled True -Action Allow -Profile ANY -Direction Outbound -Protocol TCP  -RemotePort 80
      powershell.exe New-NetFirewallRule -DisplayName 'DNS (temporary)' -Enabled True -Action Allow -Profile ANY -Direction Outbound -Protocol UDP -RemotePort 53
      powershell.exe New-NetFirewallRule -DisplayName 'DHCP (temporary)' -Enabled True -Action Allow -Profile ANY -Direction Outbound -Protocol UDP -RemotePort 67
      powershell.exe Set-NetFirewallProfile -Profile Domain,Public,Private -Enabled True
    <% end -%>

**Comment by GeoffWilliams** [comment]
@canihavethisone fix released as `0.4.0` - you were right I'd pressed close instead of merge earlier(!)

As for bulk deletion, yes its very slow on first run unfortunately. I'd say the way you fixed this is probably the right way otherwise you would end up having to figure out when to bulk purge vs rule purge. Bulk purging on a live system comes with more risk around lockouts, open firewalls while rules are set and reset connections due to firewall changes
